rust/src/passwd.rs

```diff
@@ -175,6 +175,72 @@ pub fn migrate_group_except_root(rootfs_dfd: i32, preserved_groups: &Vec<String>
     Ok(())
 }
 
+/// Recursively search a directory for a subpath owned by a UID.
+pub fn dir_contains_uid(dirfd: i32, id: u32) -> Result<bool> {
+    let dir = ffiutil::ffi_view_openat_dir(dirfd);
+    let uid = Uid::from_raw(id);
+    dir_contains_uid_gid(&dir, &Some(uid), &None)
+}
+
+/// Recursively search a directory for a subpath owned by a GID.
+pub fn dir_contains_gid(dirfd: i32, id: u32) -> Result<bool> {
+    let dir = ffiutil::ffi_view_openat_dir(dirfd);
+    let gid = Gid::from_raw(id);
+    dir_contains_uid_gid(&dir, &None, &Some(gid))
+}
+
+/// Recursively search a directory for a subpath owned by a UID or GID.
+fn dir_contains_uid_gid(dir: &openat::Dir, uid: &Option<Uid>, gid: &Option<Gid>) -> Result<bool> {
+    use openat::SimpleType;
+
+    // First check the directory itself.
+    let self_metadata = dir.self_metadata()?;
+    if compare_uid_gid(self_metadata, uid, gid) {
+        return Ok(true);
+    }
+
+    // Then recursively check all entries and subdirectories.
+    let mut found_match = false;
+    for dir_entry in dir.list_self()?.filter_map(Result::ok) {
+        let dtype = match dir_entry.simple_type() {
+            Some(t) => t,
+            None => continue,
+        };
+
+        if dtype == SimpleType::Dir {
+            let subdir_name = dir_entry.file_name();
+            if subdir_name != "." && subdir_name != ".." {
```

Comment:
I don't think those are returned by default, we don't do this check in other places.

Reply:
Self note: the `openat` library does indeed filter them internally: https://github.com/tailhook/openat/blob/eab8f97a4799f72a823fc6a8d642426f6c1fcf43/src/list.rs#L97-L98.